### ISSUE TITLE
Response gets mixed when multiple requests are executed simultaneously with filetype=csv

### ISSUE THREAD

**Issue by Madhu-NEC** [telefonicaid/fiware-sth-comet#532]
Hi,
I am using STH Comet v2.3.0 . I am facing below issue:

When multiple GET requests (http://localhost:8666/STH/v1/contextEntities/type/Room/id/Room1/attributes/temperature?lastN=300&dateFrom=2020-03-26T06:08:43.843Z&dateTo=2020-03-26T10:57:34.931Z&filetype=csv)  for each EntityID are executed simultaneously for different EntityIDs i.e Room1,Room2 and Room3.

The response of different EntityIDs gets mixed i.e. sometimes data of Room1 outputs for Room3.

**Comment by Madhu-NEC** [comment]
Hi @fgalan , @jason-fox ,
Could you please provide your feedback for below investigation?
As per my understanding, the filename created at https://github.com/telefonicaid/fiware-sth-comet/blob/master/lib/database/sthDatabase.js#L375 uses Date.now(). For same timestamp, the file gets updated before responding to the previous request that's why more than one request gets served with the same response.
To fix this, we can use uuid instead of Date.now() or maybe append uuid with the existing filename.

**Comment by Madhu-NEC** [comment]
Hi @fgalan , @jason-fox ,

Is there any update on this? I request you to please confirm.

**Comment by fgalan** [comment]
Thanks for the feedback!

I have one question regarding, please:

> When multiple GET requests (http://localhost:8666/STH/v1/contextEntities/type/Room/id/Room1/attributes/temperature?lastN=300&dateFrom=2020-03-26T06:08:43.843Z&dateTo=2020-03-26T10:57:34.931Z&filetype=csv) for each EntityID are executed simultaneously for different EntityIDs i.e Room1,Room2 and Room3.

If you do the same test but without using `&filetype=csv`, does it work as expected? I mean, the response in each one of the three request (for Room1, Room2 and Room3) is a JSON with the results only of the given entity.

Just to understand and narrow down the scope of the issue...

**Comment by Madhu-NEC** [comment]
Hi @fgalan ,

> If you do the same test but without using &filetype=csv, does it work as expected? I mean, the response in each one of the three request (for Room1, Room2 and Room3) is a JSON with the results only of the given entity.

Yes, without using **&filetype=csv** it works as expected. It is because in this case temp file is not created to read the response. The response is directly output to the request.

**Comment by Madhu-NEC** [comment]
Hi @fgalan ,

Is there any update on this? Please provide your understanding.

**Comment by fgalan** [comment]
Maybe addint the entityId to the temporal filename whould avoid the problem. Something like:

```
var fileName = entityId + '-' attrName + '-' + Date.now() + '.csv';
```

or even better:

```
var fileName = entityType + '-' + entityId + '-' + attrName + '-' + Date.now() + '.csv';
```


It seems to be an easy to do modification... maybe you could test it and if it works propose a PR contributing with the fix to the repository, please?

tThanks!

**Comment by Madhu-NEC** [comment]
Hi @fgalan 

> Maybe addint the entityId to the temporal filename whould avoid the problem. Something like:
> 
> ```
> var fileName = entityId + '-' attrName + '-' + Date.now() + '.csv';
> ```
> 
> or even better:
> 
> ```
> var fileName = entityType + '-' + entityId + '-' + attrName + '-' + Date.now() + '.csv';
> ```
> 
> It seems to be an easy to do modification... maybe you could test it and if it works propose a PR contributing with the fix to the repository, please?
> 
> tThanks!

I have tested above scenarios.
Below is the observation:
In both the cases, lets say multiple requests are executed with different parameters combinations like 
1. http://localhost:8666/STH/v1/contextEntities/type/Room/id/Room1/attributes/temperature?lastN=300&dateFrom=2020-03-26T06:08:43.843Z&dateTo=2020-03-26T10:57:34.931Z&filetype=csv
2. http://localhost:8666/STH/v1/contextEntities/type/Room/id/Room1/attributes/temperature?lastN=300&filetype=csv
3. http://localhost:8666/STH/v1/contextEntities/type/Room/id/Room1/attributes/temperature?hLimit=10&hOffset=0&filetype=csv
the results of query 1, 2 and 3 gets mixed up as the entityId and entityType is same for all of the above queries.

As all the above queries executed with `filetype=csv`, again temporary file generation may end up with same name.

So, as per my understanding, filename should be unique corresponding to each request.
Please share your understanding.

**Comment by fgalan** [comment]
Ok, let's try with a different approach :)

Maybe STH can keep an internal counter, so each time a new temporal file is generated the counter is added to the request and it's increased. That way the name of the temporal filename should be unique per request (no matter if they are for the same combination of entityid/entitytype/attribute), at least in theory.

Could you test that approach and provide feedback, please? Thanks!

**Comment by Madhu-NEC** [comment]
Hi @fgalan ,

I will test the approach mentioned [here](https://github.com/telefonicaid/fiware-sth-comet/issues/532#issuecomment-617986335) . I will let you know the results and raise the PR for the same. Thanks!

**Comment by Madhu-NEC** [comment]
Hi @fgalan ,
I have tested the approach mentioned above. The issue is resolved using above approach. I have raised the [PR](https://github.com/telefonicaid/fiware-sth-comet/pull/538) for the same. I request you to please review it.
Thanks!

**Comment by fgalan** [comment]
Fixed by PR #538
